### PR TITLE
Add Array Features to Firestore Java

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This PR adds the new array features to the Firestore Java SDK:
- FieldValue.arrayUnion()
- FieldValue.arrayRemove()
- Query.whereArrayContains()

This is not a direct port of Firestore Node, but uses the same concepts (see https://github.com/googleapis/nodejs-firestore/blob/master/dev/src/field-value.ts). API matches Firestore Android.

The .equals() methods are generated.

### CODE REVIEW

File: google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FieldValue.java [hunk 1]
```diff
@@ -16,13 +16,172 @@
...
+      }
+
+      FieldTransform.Builder fieldTransform = FieldTransform.newBuilder();
+      fieldTransform.setFieldPath(path.getEncodedPath());
+      fieldTransform.setAppendMissingElements(encodedElements);
+      return fieldTransform.build();
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (pongad) [reply google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FieldValue.java]: gah I forgot about this. This seems important.

[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FieldValue.java]: Done

File: google-cloud-clients/google-cloud-firestore/src/test/java/com/google/cloud/firestore/DocumentReferenceTest.java [hunk 2]
```diff
@@ -395,23 +404,65 @@ public void mergeWithServerTimestamps() throws Exception {
...
   @Test
-  public void serverTimestampInArray() throws Exception {
+  public void setWithArrayUnion() throws Exception {
+    doReturn(FIELD_TRANSFORM_COMMIT_RESPONSE)
+        .when(firestoreMock)
+        .sendRequest(
+            commitCapture.capture(), Matchers.<UnaryCallable<CommitRequest, CommitResponse>>any());
+
+    documentReference
+        .set(map("foo", FieldValue.arrayUnion("bar", LocalFirestoreHelper.map("foo", "baz"))))
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-clients/google-cloud-firestore/src/test/java/com/google/cloud/firestore/DocumentReferenceTest.java]: Done

**Comment by pongad** [comment]
@schmidt-sebastian Sorry I noticed this late. Now that I think about this more, I don't think calling `super.hashCode` and `super.equals` is right. For example, two `ArrayRemoveFieldValue`s are equal iff their "super instance"s are equal and the `elements` are equal. I don't think the elements will actually matter. `ArrayRemoveFieldValue::equals` calls `FieldValue::equals`, which calls `Object::equals` right? `Object::equals` only return true if the two objects are equal by reference. So, if I create `ArrayUnionFieldValue` twice with the same elements, I think the currently implementation will report false.

Am I understanding this correctly...?

**Comment by schmidt-sebastian** [comment]
> Am I understanding this correctly...?

Yeah, I am not sure why IntelliJ does this by default (or why there is no option). In the current implementation, I would only ever get false. I fixed this and added a unit test.

**Comment by pongad** [comment]
LGTM for reals now

**Pull request merged**